### PR TITLE
[MEAR-287] Fixed failure when destination directory exists

### PR DESCRIPTION
**Pull request by zzzLobster**
[MEAR-287](https://issues.apache.org/jira/browse/MEAR-287) Fixed failure when destination directory exists

### CODE REVIEW

File: src/main/java/org/apache/maven/plugins/ear/EarMojo.java [hunk 1]
```diff
@@ -443,7 +443,7 @@ private void copyModules( final JavaEEVersion javaEEVersion,
                 {
                     getLog().info( "Copying artifact [" + module + "] to [" + module.getUri() + "] (unpacked)" );
                     // Make sure that the destination is a directory to avoid plexus nasty stuff :)
-                    if ( !destinationFile.mkdirs() )
+                    if ( !destinationFile.isDirectory() && !destinationFile.mkdirs() )
```
[the remark this answers is not in the document]
Comment (mabrarov) [reply src/main/java/org/apache/maven/plugins/ear/EarMojo.java:446]: https://github.com/zzzLobster/maven-ear-plugin/pull/1

[the remark this answers is not in the document]
Comment (mabrarov) [reply src/main/java/org/apache/maven/plugins/ear/EarMojo.java:446]: Hmm... it wasn't so hard to extend the test. Now it covers both changed lines, i.e. it fails if one of changes is missing.

[the remark this answers is not in the document]
Comment (zzzLobster) [reply src/main/java/org/apache/maven/plugins/ear/EarMojo.java:446]: I have merged the Marat's PR, So the PR currently contains IT for it

File: src/main/java/org/apache/maven/plugins/ear/EarMojo.java [hunk 2]
```diff
@@ -749,7 +749,7 @@ private void changeManifestClasspath( EarModule module, File original, JavaEEVer
                 // Create a temporary work directory
                 // MEAR-167 use uri as directory to prevent merging of artifacts with the same artifactId
                 workDirectory = new File( new File( getTempFolder(), "temp" ), module.getUri() );
-                if ( workDirectory.mkdirs() )
+                if ( workDirectory.isDirectory() || workDirectory.mkdirs() )
                 {
                     getLog().debug( "Created a temporary work directory: " + workDirectory.getAbsolutePath() );
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/ear/EarMojo.java:754]: log message is now incorrect here; probably need a nested if in this case

File: src/test/java/org/apache/maven/plugins/ear/it/EarMojoIT.java [hunk 3]
```diff
@@ -993,4 +993,24 @@ public void testProject087()
     {
         doTestProject( "project-087", new String[] { "eartest-ejb-sample-one-1.0.jar", "eartest-ejb-sample-two-1.0.jar" } );
     }
+
+    /**
+     * Builds WAR and EAR as part of multi-module project twice so that the 2nd build is guaranteed to be performed when
+     * target directories and files exist.
+     * @throws Exception in case of an error.
```
Comment (elharo) [inline src/test/java/org/apache/maven/plugins/ear/it/EarMojoIT.java:1000]: this @throws clause isn't needed

Comment (mabrarov) [reply src/test/java/org/apache/maven/plugins/ear/it/EarMojoIT.java:1000]: I agree with useless of this part of this JavaDoc but I prefer to keep code consistent - all tests in this file have this `@throws` JavaDoc

File: src/main/java/org/apache/maven/plugins/ear/EarMojo.java (hunk 2 above)
Comment (mabrarov) [reply src/main/java/org/apache/maven/plugins/ear/EarMojo.java]: Logging was reworked as requested in 304c4b3

**Comment by mabrarov** [comment]
@elharo and @hboutemy,
I assume MEAR-287 is a major bug making 3.1.0 release of Maven EAR Plugin breaking existing workflows. Could we speedup / force review of this pull request and release Maven EAR Plugin 3.1.1 ASAP? 
Thank you for your work and efforts.

File: src/test/java/org/apache/maven/plugins/ear/it/EarMojoIT.java (hunk 3 above)
Comment (elharo) [reply src/test/java/org/apache/maven/plugins/ear/it/EarMojoIT.java:1000]: We don't need to be consistent with mistakes.

Comment (mabrarov) [reply src/test/java/org/apache/maven/plugins/ear/it/EarMojoIT.java:1000]: Well, do you find it appropriate if I remove this useless JavaDoc from all tests in this file?

Comment (mabrarov) [reply src/test/java/org/apache/maven/plugins/ear/it/EarMojoIT.java]: Removed in eeb458b

**Review by elharo — APPROVED** [review]

**Review by elharo — COMMENTED** [review]
Running jenkins: https://ci-builds.apache.org/job/Maven/job/maven-box/job/maven-ear-plugin/job/mear-287/

**Comment by mabrarov** [comment]
What's pending / required for merging this pull request?

**Comment by elharo** [comment]
Trying jenkins again:

https://ci-builds.apache.org/job/Maven/job/maven-box/job/maven-ear-plugin/job/17/

**Comment by jira-importer** [comment]
Resolve #452

**Comment by jira-importer** [comment]
Resolve #456

**Pull request merged**